File: perf/benchmarks/apps/levenshtein/levenshtein.go
// levenshtein — edit-distance DP application benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func makeStr(seed int64, n int) []byte {
	out := make([]byte, n)
	for i := 0; i < n; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		out[i] = byte('a') + byte((seed>>16)&7)
	}
	return out
}

func main() {
	n := 1500
	if len(os.Args) > 1 {
		v, _ := strconv.Atoi(os.Args[1])
		n = v
	}
	a := makeStr(1, n)
	b := makeStr(7, n)

	t0 := time.Now()
	prev := make([]int32, n+1)
	cur := make([]int32, n+1)
	for j := 0; j <= n; j++ {
		prev[j] = int32(j)
	}
	for i := 1; i <= n; i++ {
		cur[0] = int32(i)
		ai := a[i-1]
		for j := 1; j <= n; j++ {
			var cost int32
			if ai != b[j-1] {
				cost = 1
			}
			v := prev[j-1] + cost
			if d := cur[j-1] + 1; d < v {
				v = d
			}
			if d := prev[j] + 1; d < v {
				v = d
			}
			cur[j] = v
		}
		prev, cur = cur, prev
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", prev[n])
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/levenshtein/levenshtein.rs
// levenshtein — edit-distance DP application benchmark (Rust).
use std::env;
use std::time::Instant;

fn make_str(mut seed: i64, n: usize) -> Vec<u8> {
    let mut v = Vec::with_capacity(n);
    for _ in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        v.push(b'a' + ((seed >> 16) & 7) as u8);
    }
    v
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 { args[1].parse().unwrap_or(1500) } else { 1500 };
    let a = make_str(1, n);
    let b = make_str(7, n);

    let t0 = Instant::now();
    let mut prev: Vec<i32> = (0..=n as i32).collect();
    let mut cur: Vec<i32> = vec![0; n + 1];
    for i in 1..=n {
        cur[0] = i as i32;
        let ai = a[i - 1];
        for j in 1..=n {
            let cost = if ai == b[j - 1] { 0 } else { 1 };
            let mut v = prev[j - 1] + cost;
            let d = cur[j - 1] + 1;
            if d < v { v = d; }
            let d = prev[j] + 1;
            if d < v { v = d; }
            cur[j] = v;
        }
        std::mem::swap(&mut prev, &mut cur);
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", prev[n]);
    println!("ELAPSED_MS:{:.6}", ms);
}
